### DISCUSSION TITLE
Using the SplashScreen feature

### DISCUSSION THREAD

**Discussion by TristanPouliquen** [tauri-apps/tauri#6384] [Q&A]
Hello,

Still discovering new features of Tauri, and today I was looking into the SplashScreen one.

I wanted to display the splashscreen until I have initialized my app state (multiple queries to my DB), so I followed https://tauri.app/v1/guides/features/splashscreen/ and changed the call to the `close_splashscreen` command in the then of my `Promise.all([<various invoke calls>])`.

However my splashscreen does not get closed, so I assume my initializing state function is not called. It is called in a `useEffect`.

I suppose the problem might come from there, in that the `useEffect` is called when my main component from my homepage is mounted, and it is not while the main window is hidden.

If I call my initializing method outside of a `useEffect`, I get an error stating that `window is not defined`.

Adding the code snippet with `document.addEventListener` fails too as the component needs to be mounted to be able to access `document`.

Does someone know of a way I haven't thought of that could make this work (closing the splashScreen after some promises are resolved) or is my usecase just impossible?

**Comment by FabianLars** [comment]
Your splashscreen branch seems to work fine for me on second try. Like, it seemingly didn't work on the first try but since then it never failed and i can't reproduce it anymore 😅

**Reply by TristanPouliquen** [reply]
Hum, when I bundle it, it indeed works, but in dev mode, it stays stuck on the splash screen. In which case have you tried?

If it's only in dev mode, it will work but it's going to be difficult to continue development once I merge this branch 😅

**Reply by TristanPouliquen** [reply]
Whil running `yarn tauri dev`, I get this output: 

```
tristan@tristan-assoconnect:~/dev/perso/selection-poste-enm$ yarn tauri dev
     Running BeforeDevCommand (`npm run dev`)

> selection-poste-enm@0.2.0 dev
> next dev

ready - started server on 0.0.0.0:3000, url: http://localhost:3000

🌼 daisyUI components 2.51.3  https://daisyui.com
        Info Watching /home/tristan/dev/perso/selection-poste-enm/src-tauri for changes...
  ✔︎ Including:  base, components, 29 themes, utilities
  ❤︎ Support daisyUI:  https://opencollective.com/daisyui 
  
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
event - compiled client and server successfully in 1555 ms (169 modules)
wait  - compiling...
event - compiled successfully in 181 ms (135 modules)
wait  - compiling...
event - compiled successfully in 8 ms (34 modules)
wait  - compiling / (client and server)...
event - compiled client and server successfully in 1583 ms (1484 modules)
```

My frontend code seems to compile multiple times without reason, maybe something is going on there. Do you have any ideas as to how I could get info to understand what is going on?

**Reply by FabianLars** [reply]
> In which case have you tried?

Only dev so far.

> My frontend code seems to compile multiple times without reason, maybe something is going on there. Do you have any ideas as to how I could get info to understand what is going on?

Nope, frontend stuff is mostly out of my league.

Did you try adding a setTimeout (or equivalent in rust) to force-show the window after x seconds? Maybe the devtools console shows something useful.

**Reply by TristanPouliquen** [reply]
Hi, I've had more time to test out things yesterday! 

I added debug info to my Tauri command that should be called & close the splashscreen.

Without forcing a timeout in my app setup (following https://tauri.app/v1/guides/features/splashscreen/#waiting-for-rust), this command does not get called.

If I force a timeout in Rust, my splashscreen closes correctly, my main window opens & then my command to close the splashscreen is called in turn (but at this point, it only finds the "main" window in any case).

I still do not understand why & how you managed to have it working once in dev, and why I have it working when building the executables 🤔 
As is, the main window javascript does not seem to be evaluated & executed, which prevents events coming from it to be fired.

**Reply by TristanPouliquen** [reply]
Hello, would you know how we could advance on this issue? 

Is this feature working for you on other projects (ie: is this issue linked to something in our project)? This doesn't seem plausible as apparently our JS is never invoked at all but I may have missed something.

How could we ascertain that the main window is effectively opened & JS loaded in it to begin JS data fetching?

**Reply by TristanPouliquen** [reply]
I tried building a new Tauri App to see if the problem is coming from my code or not, but after running `yarn create tauri-app`, choosing Javascript, React, Typescript as my frontend options, and getting the template with VITE, when I run `yarn tauri dev`, my console crashes :/ 

I'll try an install following the NextJS guide I had followed in the beginning.

**Reply by TristanPouliquen** [reply]
By following https://tauri.app/fr/v1/guides/getting-started/setup/next-js/ to start a NextJS project from scratch, and adding code from https://tauri.app/v1/guides/features/splashscreen/#docusaurus_skipToContent_fallback to have my JS code call the tauri command in a `useEffect` with a timeout of 2 seconds, I get stuck on the splashscreen too.

**Reply by FabianLars** [reply]
Can you share the output of `tauri info` again? Or actually just your OS + version.

**Reply by TristanPouliquen** [reply]
Hi, here is the output of `tauri info` :

```
Environment
  › OS: Ubuntu 22.04 X64
  › Node.js: 18.13.0
  › npm: 8.19.3
  › pnpm: Not installed!
  › yarn: 3.4.1
  › rustup: 1.25.2
  › rustc: 1.67.0
  › cargo: 1.67.0
  › Rust toolchain: stable-x86_64-unknown-linux-gnu 

Packages
  › @tauri-apps/cli [NPM]: 1.2.3
  › @tauri-apps/api [NPM]: 1.2.0
  › tauri [RUST]: 1.2.4,
  › tauri-build [RUST]: 1.2.1,
  › tao [RUST]: 0.15.8,
  › wry [RUST]: 0.23.4,

App
  › build-type: bundle
  › CSP: unset
  › distDir: ../out
  › devPath: http://localhost:3000/
  › framework: React (Next.js)
  › bundler: Webpack

App directory structure
  ├─ _state
  ├─ src-tauri
  ├─ out
  ├─ .next
  ├─ components
  ├─ .yarn
  ├─ .husky
  ├─ .github
  ├─ .idea
  ├─ node_modules
  ├─ public
  ├─ styles
  ├─ types
  ├─ pages
  └─ .git
```

**Reply by FabianLars** [reply]
Thanks. Is that default Ubuntu with the Gnome DE?

**Reply by FabianLars** [reply]
Interesting, on Ubuntu with Nextjs i can indeed reproduce it a bit more reliable. Seems like it indeed doesn't mount the component in invisible windows :/
The only thing that worked was this _outside_ of a components function which makes it kinda useless ngl
```ts
if (typeof window !== 'undefined') {
    invoke('close_splashscreen').then(console.log).catch(console.error);
}
```

**Reply by FabianLars** [reply]
Seems to work fine after `tauri build` for me though, which i guess is _something_...

**Reply by TristanPouliquen** [reply]
Yes, when building, the splashscreen works fine, it is just in dev mode that I get this issue, which is weird. 

But I don't want to merge this feature in the main branch as is, as it will make it near impossible to easily dev next features 😅

**Reply by FabianLars** [reply]
the only thing i can think of is to just call `window.show()` in rust annotated with `#[cfg(dev)]`, because i have no idea how to tell nextjs that it's fine to mount the components even if the window/document is hidden x)

**Reply by TristanPouliquen** [reply]
Where would I add this rust annotation?

**Reply by FabianLars** [reply]
for example in a setup hook like this
```rs
tauri::Builder::default()
  .setup(|app| {
// If you already have a setup hook which requests the window you could only annotate the win.show() line instead of this whole block
    #[cfg(dev)]
    {
      use tauri::Manager;
      let win = app.get_window("main").unwrap();
      let _ = win.show();
    }
    Ok(())
  })
  ...
```
